src/app/zeko/da_layer/lib/node.ml

```diff
@@ -106,32 +128,129 @@ let post_batch t ~ledger_openings ~batch =
             else
               Error
-                (Error.create "Index mismatch" index Account.Index.sexp_of_t)
+                (Error.of_string
+                   (sprintf "Index mismatch %d <> %d" opening_index diff_index) )
           in
-          Ok (Sparse_ledger.set_exn ledger index account)
+          Ok (Sparse_ledger.set_exn ledger diff_index account)
         with e -> Error (Error.of_exn e) )
   in
-  let target_ledger_hash = Sparse_ledger.merkle_root target_ledger in
+  let target_ledger_hash = Batch.target_ledger_hash batch in
+  let%bind.Result () =
+    if
+      Ledger_hash.equal target_ledger_hash
+        (Sparse_ledger.merkle_root target_ledger)
+    then Ok ()
+    else
+      Error
+        (Error.create "Target ledger hash mismatch"
+           (target_ledger_hash, Sparse_ledger.merkle_root target_ledger)
+           [%sexp_of: Ledger_hash.t * Ledger_hash.t] )
+  in
+
+  (* 5 *)
   let message =
     Random_oracle.Input.Chunked.field_elements [| target_ledger_hash |]
   in
   let signature = Schnorr.Chunked.sign t.signer.private_key message in
 
-  (* 4 *)
+  (* 6 *)
+  let get_account ledger account_id =
+    try
+      let index = Sparse_ledger.find_index_exn ledger account_id in
+      Ok (Sparse_ledger.get_exn ledger index)
+    with e -> Error (Error.of_exn e)
+  in
+  (* First try to find in [map] and fallback to the [ledger_openings] *)
+  let get_account's_receipt_chain_hash map account_id =
+    match Account_id.Map.find map account_id with
+    | Some account ->
+        Ok account
+    | None ->
+        let%bind.Result account = get_account ledger_openings account_id in
+        Ok account.receipt_chain_hash
+  in
+  let%bind.Result applied_hashes =
+    match Batch.command_with_action_step_flags batch with
+    | None ->
+        Ok Account_id.Map.empty
+    | Some (Signed_command command, _) ->
+        (* For command only the fee payer gets the receipt *)
+        let account_id = Signed_command.fee_payer command in
+        let%bind.Result old_receipt_chain_hash =
+          get_account's_receipt_chain_hash Account_id.Map.empty account_id
+        in
+        let new_receipt_chain_hash =
+          Receipt.Chain_hash.cons_signed_command_payload
+            (Signed_command_payload (Signed_command.payload command))
+            old_receipt_chain_hash
+        in
+        Ok
+          (Account_id.Map.set Account_id.Map.empty ~key:account_id
+             ~data:new_receipt_chain_hash )
+    | Some (Zkapp_command command, _) ->
+        let _commitment, full_transaction_commitment =
+          Zkapp_command.get_transaction_commitments command
+        in
+        let%bind.Result _, acc =
+          List.fold_result (Zkapp_command.all_account_updates_list command)
+            ~init:(Unsigned.UInt32.zero, Account_id.Map.empty)
+            ~f:(fun (index, acc) account_update ->
+              (* Receipt chain hash is updated only for account updates authorised with Proof of Signature *)
```

Comment:
I didn’t know this was the case lol

Reply:
It makes sense that you can't mess with the receipt chain hash of someone else's account.